zebrad/src/components/inbound.rs

```diff
@@ -110,96 +131,131 @@ impl Service<zn::Request> for Inbound {
-                        Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
-                        Timeout::new(self.verifier.clone(), BLOCK_VERIFY_TIMEOUT),
-                        self.state.clone(),
-                    )));
-                }
-                Err(TryRecvError::Empty) => {
-                    self.network_setup = Some(rx);
+        if matches!(self.network, Setup::AwaitingNetwork { .. }) {
+            // Unfortunately, we can't match, swap, and destructure at the same time
+            let mut awaiting_state = Setup::FailedInit;
+            mem::swap(&mut self.network, &mut awaiting_state);
+            if let Setup::AwaitingNetwork {
+                mut network_setup,
+                verifier,
+            } = awaiting_state
+            {
+                match network_setup.try_recv() {
+                    Ok((outbound, address_book)) => {
+                        let downloads = Box::pin(Downloads::new(
+                            Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
+                            Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT),
+                            self.state.clone(),
+                        ));
+                        self.network = Setup::Initialized {
+                            address_book,
+                            downloads,
+                        };
+                    }
+                    Err(TryRecvError::Empty) => {
+                        // There's no setup data yet, so keep waiting for it
+                        self.network = Setup::AwaitingNetwork {
+                            network_setup,
+                            verifier,
+                        };
+                    }
+                    Err(error @ TryRecvError::Closed) => {
+                        // Mark the service as failed, because network setup failed
+                        error!(?error, "inbound network setup failed");
+                        self.network = Setup::FailedRecv {
+                            failed_setup: network_setup,
+                        };
+                        return Poll::Ready(Err(error.into()));
+                    }
                 }
-                Err(e @ TryRecvError::Closed) => {
-                    // In this case, report that the service failed, and put the
-                    // failed oneshot back so we'll fail again in case
-                    // poll_ready is called after failure.
-                    self.network_setup = Some(rx);
-                    return Poll::Ready(Err(e.into()));
-                }
-            };
+            }
         }
 
-        // Clean up completed download tasks
-        if let Some(downloads) = self.downloads.as_mut() {
-            while let Poll::Ready(Some(_)) = downloads.as_mut().poll_next(cx) {}
+        // Unfortunately, we can't combine these matches into an exhaustive match statement,
+        // because they use mutable references, or they depend on the state we've just modified.
```

Comment:
I think this shouldn't be the case anymore once we've introduced the `take_setup` method I mentioned above.